torch/nn/functional.py

```diff
@@ -30,13 +30,22 @@
     {cudnn_reproducibility_note}
 """.format(**reproducibility_notes, **tf32_notes) + r"""
 
+    Padding mode 'valid' is the same as no padding. 'same' pads the input so
+    the output has shape ``ceil(in_shape / stride)`` in each dimension.
+    For the default ``stride = 1``, the output is the same size as the input.
+
+    .. warning::
+        For ``padding='same'``, if the ``weight`` is even-length and
+        ``dilation`` is odd in any dimension, a full :func:`pad` operation
+        may be needed internally. Lowering performance.
```

Comment:
nit: "may be needed internally. Lowering performance." -> "will be needed internally, lowering performance."

Is "even-length weight" & "odd dilation" the only case that requires a full pad operation? If so, please update the `TORCH_WARN_ONCE` text to specify this.

Reply:
mkldnn supports asymmetric padding, so whether or not it happens is an implementation detail. See next PR in the stack which takes advantage of that.